exercises/concept/the-farm/the_farm_test.go

```diff
@@ -14,192 +14,141 @@ func (wf testWeightFodder) FodderAmount() (float64, error) {
-		t.Run(test.description, func(t *testing.T) {
-			gotAmount, gotErr := DivideFood(test.weightFodder, test.cows)
-			switch {
-			case gotAmount != test.wantAmount:
-				t.Errorf(
-					"DivideFood(weightFodder(%v), %v) got (%v, %v) wanted (%v, %v)",
-					test.weightFodderDescription,
-					test.cows,
-					gotAmount,
-					gotErr,
-					test.wantAmount,
-					test.wantErr,
-				)
-			case gotErr != nil && test.wantErr == nil:
-				t.Errorf(
-					"DivideFood(weightFodder(%v), %v) got an unexpected error (%v)",
-					test.weightFodderDescription,
-					test.cows,
-					gotErr,
-				)
-			case gotErr == nil && test.wantErr != nil:
-				t.Errorf(
-					"DivideFood(weightFodder(%v), %v) got no error, but wanted an error (%v)",
-					test.weightFodderDescription,
-					test.cows,
-					test.wantErr,
-				)
-			case !(gotErr == test.wantErr || gotErr.Error() == test.wantErr.Error()):
-				t.Errorf(
-					"DivideFood(weightFodder(%v), %v) got error (%v), but wanted error (%v)",
-					test.weightFodderDescription,
-					test.cows,
-					gotErr,
-					test.wantErr,
-				)
-			}
-		})
+		t.Logf("Testing condition %s", test.description)
+
+		gotFodderAmount, gotErr := DivideFood(test.fodderAmount, test.cows)
+
+		if gotFodderAmount != test.wantFodderAmount && gotErr != test.wantFodderErr {
+			t.Errorf("Wrong food division ... \ngot %v and %v, \nwant %v and %v",
+				gotFodderAmount, gotErr, test.wantFodderAmount, test.wantFodderErr)
+		}
 	}
 }
 
 func TestDivideFoodSillyNephewError(t *testing.T) {
 	tests := []struct {
 		description string
 		cows        int
-		wantErrMsg  string
+		wantErrMsg  error
 	}{
 		{
-			description: "Negative ten cows",
+			description: "negative ten cows",
 			cows:        -10,
-			wantErrMsg:  "silly nephew, there cannot be -10 cows",
+			wantErrMsg:  errors.New("silly nephew, there cannot be -10 cows"),
 		},
 		{
-			description: "Negative seven cows",
+			description: "negative seven cows",
 			cows:        -7,
-			wantErrMsg:  "silly nephew, there cannot be -7 cows",
+			wantErrMsg:  errors.New("silly nephew, there cannot be -7 cows"),
 		},
 	}
+
 	for _, test := range tests {
-		t.Run(test.description, func(t *testing.T) {
-			weightFodder := testWeightFodder{fodder: 100, err: nil}
-			gotAmount, gotErr := DivideFood(weightFodder, test.cows)
-			if gotAmount != 0 {
-				t.Errorf(
-					"DivideFood(weightFodder(%v), %v) got amount %v, but wanted amount 0",
-					"100 fodder, no error",
-					test.cows,
-					gotAmount,
-				)
-			}
-			if gotErr.Error() != test.wantErrMsg {
-				t.Errorf(
-					"DivideFood(weightFodder(%v), %v) got error msg %q, but wanted error msg %q",
-					"100 fodder, no error",
-					test.cows,
-					gotErr.Error(),
-					test.wantErrMsg,
-				)
-			}
-		})
+		t.Logf("Testing for %s", test.description)
+
+		fodderWeight := testWeightFodder{fodder: 100, err: nil}
+		gotFodderAmount, gotErrMsg := DivideFood(fodderWeight, test.cows)
+
+		if gotFodderAmount != 0 && gotErrMsg != test.wantErrMsg {
```

Comment:
Errors cannot be compared in this way, see https://go.dev/play/p/rRmtE4fDdQ9
Also, we don't expect a normal sentinel error here, we expect a special type of error that the student should write.

Reply:
The special type of error that the student should write is already a part of `DivideFood`. This method returns the amount of fodder and error message. A combination of these return values are being compared with expected values from test data.